plugins/http_plugin/http_plugin.cpp

```diff
@@ -123,22 +163,132 @@ namespace eosio {
-            auto resolved_port_str = std::to_string(ep.port());
-            plugin_state->valid_hosts.emplace(host + ":" + port);
-            plugin_state->valid_hosts.emplace(host + ":" + resolved_port_str);
+         bool on_loopback_only(const std::string& address) {
+            if (is_unix_socket_address(address))
+               return true;
+            auto [host, port] = split_host_port(address);
+            boost::system::error_code ec;
+            tcp::resolver             resolver(plugin_state->thread_pool.get_executor());
+            auto endpoints = resolver.resolve(host, port, boost::asio::ip::tcp::resolver::passive, ec);
+            EOS_ASSERT(!ec, chain::plugin_config_exception, "failed to resolve address: ${msg}", ("msg", ec.message()));
+            return std::all_of(endpoints.begin(), endpoints.end(), [](const auto& ep) {
+               return ep.endpoint().address().is_loopback();
+            });
          }
 
-         void create_beast_server(bool isUnix) {
-            if(isUnix) {
-               beast_unix_server = std::make_shared<beast_http_listener<unix_socket_session, stream_protocol, stream_protocol::socket> >(plugin_state);
-               fc_ilog( logger(), "created beast UNIX socket listener");
+         void create_beast_server(const std::string& address, api_category_set categories) {
+
+            EOS_ASSERT(address.size() >= 2, chain::plugin_config_exception, "Invalid http server address: ${addr}",
+                       ("addr", address));
+
+            try {
+               if (is_unix_socket_address(address)) {
+                  auto                  cwd       = std::filesystem::current_path();
+                  std::filesystem::path sock_path = address;
+                  if (sock_path.is_relative())
+                     sock_path = std::filesystem::weakly_canonical(app().data_dir() / sock_path);
+                  std::filesystem::remove(sock_path);
+                  std::filesystem::create_directories(sock_path.parent_path());
+                  // The maximum length of the socket path is defined by sockaddr_un::sun_path. On Linux,
+                  // according to unix(7), it is 108 bytes. On FreeBSD, according to unix(4), it is 104 bytes.
+                  // Therefore, we create the unix socket with the relative path to its parent path to avoid the problem.
+                  std::filesystem::current_path(sock_path.parent_path());
+                  auto restore = fc::make_scoped_exit([cwd]{
+                     std::filesystem::current_path(cwd);
+                  });
+
+                  using stream_protocol = asio::local::stream_protocol;
+                  auto server = std::make_shared<beast_http_listener<stream_protocol::socket>>(
+                        plugin_state, categories, stream_protocol::endpoint{ sock_path.filename().string() });
+                  server->do_accept();
```

Comment:
In the previous version of http_plugin, when we created `beast_http_listener`, we stored the shared_ptr to these, so we could call `stop_listening` and delete the server in `plugin_shutdown`. Why don't we do that anymore?

Reply:
@heifner what do you think about this. I still think that explicitly shutting down the http servers in `plugin_shutdown` is cleaner, but if you believe it is equivalent I am fine to go with the simplest solution.

Reply:
As is I believe is the pattern we want to move to. Just allow it to naturally shutdown.